state-sync/diem-data-client/src/lib.rs

```diff
@@ -0,0 +1,158 @@
+
+/// A response error that users of the Diem Data Client can use to notify
+/// the Data Client about invalid or malformed responses.
+#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
+pub enum ResponseError {
+    InvalidData,
+    MissingData,
+    ProofVerificationError,
+}
+
+/// The API offered by the Diem Data Client.
+#[async_trait]
+pub trait DiemDataClient {
+    /// Returns a single account states chunk with proof, containing the accounts
+    /// from start to end index (inclusive) at the specified version. The proof
+    /// version is the same as the specified version.
+    async fn get_account_states_with_proof(
+        &self,
+        version: u64,
+        start_index: HashValue,
+        end_index: HashValue,
+    ) -> Result<DataClientResponse, Error>;
+
+    /// Returns all epoch ending ledger infos between start and end (inclusive).
+    /// If the data cannot be fetched (e.g., the number of epochs is too large),
+    /// an error is returned.
+    async fn get_epoch_ending_ledger_infos(
+        &self,
+        start_epoch: u64,
+        end_epoch: u64,
+    ) -> Result<DataClientResponse, Error>;
+
+    /// Returns a global summary of the data currently available in the network.
+    fn get_global_data_summary(&self) -> Result<DataClientResponse, Error>;
+
+    /// Returns the number of account states at the specified version.
+    async fn get_number_of_account_states(&self, version: u64)
+        -> Result<DataClientResponse, Error>;
+
+    /// Returns a transaction output list with proof object, with transaction
+    /// outputs from start to end versions (inclusive). The proof is relative to
+    /// the specified `proof_version`. If the data cannot be fetched (e.g., the
+    /// number of transaction outputs is too large), an error is returned.
+    async fn get_transaction_outputs_with_proof(
+        &self,
+        proof_version: u64,
+        start_version: u64,
+        end_version: u64,
+    ) -> Result<DataClientResponse, Error>;
+
+    /// Returns a transaction list with proof object, with transactions from
+    /// start to end versions (inclusive). The proof is relative to the specified
+    /// `proof_version`. If the data cannot be fetched (e.g., the number of
+    /// transactions is too large), an error is returned.
+    async fn get_transactions_with_proof(
+        &self,
+        proof_version: u64,
+        start_version: u64,
+        end_version: u64,
+    ) -> Result<DataClientResponse, Error>;
+
+    /// Notifies the Diem Data Client about a previously received response that
+    /// was bad (e.g., invalid or malformed).
+    async fn notify_bad_response(
+        &self,
+        response_id: u64,
+        response_error: ResponseError,
+    ) -> Result<(), Error>;
+}
+
+/// A response from the Data Client for a single API call. The `response_id` can
+/// be used to provide feedback to the Data Client (e.g., invalid chunks).
+pub struct DataClientResponse {
+    pub response_id: u64,
```

Comment:
same, why not have request id within the response? also are we writing this in such a way that it can run on any transport, as opposed to json-rpc or our rpc that already has concepts for ids defined?

Reply:
- Regarding request ids, they're not really needed. Users of `DiemDataClient` will call this API and await a response, so there's no need to use request ids -- we know the response once the call returns. Not sure if I've missed something.
- Regarding the transport layer, that's probably a question for Philip, but my